src/slurm_plugin/clustermgtd.py

```diff
@@ -853,47 +722,268 @@ def _handle_unhealthy_static_nodes(self, unhealthy_static_nodes, private_ip_to_i
+                    # ignore nodes not belonging to any partition
+                    ignored_nodes.append(node)
+                else:
+                    for p in node.partitions:
+                        partitions_name_map[p].slurm_nodes.append(node)
+            if ignored_nodes:
+                log.warning("Ignoring following nodes because they do not belong to any partition: %s", ignored_nodes)
+            return partitions_name_map
+        except Exception as e:
+            log.error("Failed when getting partition/node states from scheduler with exception %s", e)
+            raise
+
+    @staticmethod
+    def _update_slurm_nodes_with_ec2_info(nodes, cluster_instances):
+        if cluster_instances:
+            ip_to_slurm_node_map = {node.nodeaddr: node for node in nodes}
+            for instance in cluster_instances:
+                if instance.private_ip in ip_to_slurm_node_map:
+                    slurm_node = ip_to_slurm_node_map.get(instance.private_ip)
+                    slurm_node.instance = instance
+                    instance.slurm_node = slurm_node
+
+    @staticmethod
+    def get_instance_id_to_active_node_map(partitions):
+        instance_id_to_active_node_map = {}
+        for partition in partitions:
+            if not partition.is_inactive():
+                for node in partition.slurm_nodes:
+                    if node.instance:
+                        instance_id_to_active_node_map[node.instance.id] = node
+        return instance_id_to_active_node_map
+
+    @staticmethod
+    def _get_inactive_instances_and_nodes(partitions):
+        inactive_instance_ids = set()
+        inactive_nodes = set()
+        try:
+            for partition in partitions:
+                if partition.is_inactive():
+                    inactive_nodes |= set(partition.slurm_nodes)
+                    inactive_instance_ids |= {node.instance.id for node in partition.slurm_nodes if node.instance}
+        except Exception as e:
+            log.error("Unable to get inactive instances and nodes. Exception: %s", e)
+        return inactive_instance_ids, inactive_nodes
+
+    def _is_node_being_replaced(self, node):
+        """Check if a node is currently being replaced and within node_replacement_timeout."""
+        if (
+            node.instance
+            and node.name in self._static_nodes_in_replacement
+            and not time_is_up(
+                node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
+            )
+        ):
+            return True
+        return False
+
+    def _is_node_replacement_timeout(self, node):
+        if (
+            node.instance
+            and node.name in self._static_nodes_in_replacement
+            and time_is_up(
+                node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
+            )
+        ):
+            return True
+        return False
```

Comment:
These 2 functions look like duplicates

Reply:
I thought about this, if I want to reduce duplicate, we will need to add a parameter “check_is_valid” to merge them into a single method.

It will become:
```
def _is_node_in_replacement_valid(self, node, check_is_valid):
    """Check if a node is currently being replaced and within node_replacement_timeout."""
    if (node.instance and node.name in self._static_nodes_in_replacement)
	if (check_is_valid)
		return not time_is_up(
            		node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
        	)
	else
		return time_is_up(
            		node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
        	)	
        )
    ):
        return True
    return False
```
The problem is:
* we still got duplicate time_is_up here
* the `check_is_valid` parameter is not easy to understand when people want to use it

This can be further simplified into: (by using XOR to avoid duplicating time_is_up, which makes it even more difficult to understand)

```
def _is_node_in_replacement_valid(self, node, check_is_valid):
    """Check if a node is currently being replaced and within node_replacement_timeout."""
    if (
        node.instance
        and node.name in self._static_nodes_in_replacement
        and (check_is_valid xor time_is_up(
            node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
        ))
    ):
        return True
    return False
```
So I preferred using two methods improve code readability. I could do:
```
def _is_node_with_instance_in_replacement(self, node):
        if node.instance and node.name in self._static_nodes_in_replacement:
            return True
        return False

def _is_node_being_replaced(self, node):
    """Check if a node is currently being replaced and within node_replacement_timeout."""
    if self._is_node_with_instance_in_replacement(node) and not time_is_up(
        node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
    ):
        return True
    return False

def _is_node_replacement_timeout(self, node):
        if self._is_node_with_instance_in_replacement(node)
            and time_is_up(
                node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
            )
        ):
            return True
        return False
```
Any idea on this?

Reply:
Change it to a better structure after discuss with Mauri
```    
def _is_node_being_replaced(self, node):
        """Check if a node is currently being replaced and within node_replacement_timeout."""
        return self._is_node_in_replacement_valid(node, check_node_is_valid=True)
def _is_node_replacement_timeout(self, node):
  """Check if a static node is in replacement but replacement time is expired."""
  return self._is_node_in_replacement_valid(node, check_node_is_valid=False)

def _is_node_in_replacement_valid(self, node, check_node_is_valid):
        """
  Check node is replacement timeout or in replacement.

  If check_node_is_valid=True, check whether a node is in replacement,
  If check_node_is_valid=False, check whether a node is replacement timeout.
  """
  if node.instance and node.name in self._static_nodes_in_replacement:
      time_is_expired = time_is_up(
          node.instance.launch_time, self._current_time, grace_time=self._config.node_replacement_timeout
      )
      return not time_is_expired if check_node_is_valid else time_is_expired
  return False
  ```